flocker/node/agents/blockdevice.py

```diff
@@ -55,6 +58,70 @@
 PROFILE_METADATA_KEY = u"clusterhq:flocker:profile"
 
 
+class DatasetStates(Names):
+    """
+    States that a ``Dataset`` can be in.
+    """
+    NON_MANIFEST = NamedConstant()
+    ATTACHED = NamedConstant()
+    MOUNTED = NamedConstant()
+    DELETED = NamedConstant()
+
+
+class DiscoveredDataset(PClass):
+    """
+    Dataset as discovered by deployer.
+
+    :ivar DatasetStates state: The state this dataset was determined to be in.
+    :ivar int maximum_size: The maximum size of the dataset.
+    :param unicode blockdevice_id: The unique identifier of the
+        ``IBlockDeviceAPI``-managed volume.
+    :ivar FilePath device_path: The absolute path to the block device file on
+        the node where the dataset is attached.
+    :ivar FilePath mount_point: The absolute path to the location on the node
+        where the dataset will be mounted.
+    """
+    state = field(
+        invariant=lambda state: (state in DatasetStates.iterconstants(),
+                                 "Not a valid state"),
+        mandatory=True,
+    )
+    dataset_id = field(type=UUID, mandatory=True)
+    maximum_size = field(type=int, mandatory=True)
+    blockdevice_id = field(type=unicode, mandatory=True)
+    device_path = field(FilePath)
+    mount_point = field(FilePath)
```

Comment:
+1 for raising an exception on access rather than manipulating None as a FilePath. Sort of seems like we could have different PClasses for datasets in different states.... but I think I like your approach better.